File: backend/email_service.py
"""
Email Service - Brevo (SendinBlue) Integration
Handles all email sending functionality
"""

import os
from typing import List, Optional
import requests
from datetime import datetime

BREVO_API_KEY = os.getenv("BREVO_API_KEY")
BREVO_API_URL = "https://api.brevo.com/v3/smtp/email"
FROM_EMAIL = os.getenv("FROM_EMAIL", "noreply@xtyl.com")
FROM_NAME = os.getenv("FROM_NAME", "XTYL Creativity Machine")
FRONTEND_URL = os.getenv("FRONTEND_URL", "http://localhost:3000")


def send_email(
    to_email: str,
    subject: str,
    html_content: str,
    to_name: Optional[str] = None
) -> bool:
    """
    Send email using Brevo API

    Args:
        to_email: Recipient email address
        subject: Email subject
        html_content: HTML content of the email
        to_name: Recipient name (optional)

    Returns:
        bool: True if email was sent successfully, False otherwise
    """
    if not BREVO_API_KEY:
        print("⚠️ BREVO_API_KEY not configured. Email not sent.")
        print(f"Would send to: {to_email}")
        print(f"Subject: {subject}")
        return False

    headers = {
        "accept": "application/json",
        "api-key": BREVO_API_KEY,
        "content-type": "application/json"
    }

    payload = {
        "sender": {
            "name": FROM_NAME,
            "email": FROM_EMAIL
        },
        "to": [
            {
                "email": to_email,
                "name": to_name or to_email.split("@")[0]
            }
        ],
        "subject": subject,
        "htmlContent": html_content
    }

    try:
        response = requests.post(BREVO_API_URL, json=payload, headers=headers)
        response.raise_for_status()
        print(f"✓ Email sent to {to_email}: {subject}")
        return True
    except requests.exceptions.RequestException as e:
        print(f"✗ Failed to send email to {to_email}: {e}")
        if hasattr(e, 'response') and e.response is not None:
            print(f"Response: {e.response.text}")
        return False


def send_password_reset_email(email: str, reset_token: str, user_name: Optional[str] = None) -> bool:
    """
    Send password reset email

    Args:
        email: User's email address
        reset_token: Password reset token
        user_name: User's name (optional)

    Returns:
        bool: True if email was sent successfully
    """
    reset_link = f"{FRONTEND_URL}/reset-password?token={reset_token}"

    html_content = f"""
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Recuperação de Senha - XTYL</title>
    </head>
    <body style="font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif; line-height: 1.6; color: #333; max-width: 600px; margin: 0 auto; padding: 20px;">
        <div style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); padding: 30px; text-align: center; border-radius: 10px 10px 0 0;">
            <h1 style="color: white; margin: 0; font-size: 28px;">🔐 Recuperação de Senha</h1>
        </div>

        <div style="background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px;">
            <p style="font-size: 16px; margin-bottom: 20px;">
                Olá{f" {user_name}" if user_name else ""},
            </p>

            <p style="font-size: 16px; margin-bottom: 20px;">
                Você solicitou a recuperação de senha para sua conta no <strong>XTYL Creativity Machine</strong>.
            </p>

            <p style="font-size: 16px; margin-bottom: 30px;">
                Clique no botão abaixo para criar uma nova senha:
            </p>

            <div style="text-align: center; margin: 30px 0;">
                <a href="{reset_link}"
                   style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
                          color: white;
                          padding: 15px 40px;
                          text-decoration: none;
                          border-radius: 5px;
                          font-weight: bold;
                          display: inline-block;
                          font-size: 16px;">
                    Redefinir Senha
                </a>
            </div>

            <p style="font-size: 14px; color: #666; margin-top: 30px;">
                Ou copie e cole este link no seu navegador:
            </p>
            <p style="font-size: 12px; color: #667eea; word-break: break-all; background: white; padding: 10px; border-radius: 5px;">
                {reset_link}
            </p>

            <p style="font-size: 14px; color: #666; margin-top: 30px; padding-top: 20px; border-top: 1px solid #ddd;">
                ⏰ Este link expira em <strong>1 hora</strong>.<br>
                ⚠️ Se você não solicitou esta recuperação, ignore este email.
            </p>
        </div>

        <div style="text-align: center; margin-top: 20px; color: #999; font-size: 12px;">
            <p>© {datetime.now().year} XTYL Creativity Machine. Todos os direitos reservados.</p>
        </div>
    </body>
    </html>
    """

    return send_email(
        to_email=email,
        subject="🔐 Recuperação de Senha - XTYL",
        html_content=html_content,
        to_name=user_name
    )


def send_workspace_invite_email(email: str, workspace_name: str, invited_by: str, invite_token: str, temp_password: str = None) -> bool:
    """
    Send workspace invitation email to non-registered users

    Args:
        email: Recipient's email address
        workspace_name: Name of the workspace
        invited_by: Name of the person who invited
        invite_token: Invitation token for signup
        temp_password: Temporary password for the new account

    Returns:
        bool: True if email was sent successfully
    """
    signup_link = f"{FRONTEND_URL}/signup?invite={invite_token}&email={email}"

    html_content = f"""
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Convite para XTYL</title>
    </head>
    <body style="font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif; line-height: 1.6; color: #333; max-width: 600px; margin: 0 auto; padding: 20px;">
        <div style="background: linear-gradient(135deg, #5B8DEF 0%, #4A7AD9 100%); padding: 30px; text-align: center; border-radius: 10px 10px 0 0;">
            <h1 style="color: white; margin: 0; font-size: 28px;">✉️ Você foi convidado!</h1>
        </div>

        <div style="background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px;">
            <p style="font-size: 16px; margin-bottom: 20px;">
                Olá,
            </p>

            <p style="font-size: 16px; margin-bottom: 20px;">
                <strong>{invited_by}</strong> convidou você para participar do workspace <strong>"{workspace_name}"</strong> no <strong>XTYL Creativity Machine</strong>.
            </p>

            <p style="font-size: 16px; margin-bottom: 30px;">
                O XTYL é uma plataforma completa para criação e gerenciamento de conteúdo criativo com inteligência artificial.
            </p>

            <div style="background: white; padding: 20px; border-radius: 5px; margin: 20px 0;">
                <h3 style="margin-top: 0; color: #5B8DEF;">✨ O que você pode fazer:</h3>
                <ul style="padding-left: 20px;">
                    <li style="margin-bottom: 10px;">📝 Criar e gerenciar documentos com IA</li>
                    <li style="margin-bottom: 10px;">🎨 Gerar imagens com modelos avançados</li>
                    <li style="margin-bottom: 10px;">💬 Chat inteligente com contexto</li>
                    <li style="margin-bottom: 10px;">📊 Organizar projetos em pastas</li>
                    <li style="margin-bottom: 10px;">🖼️ Biblioteca de assets visuais</li>
                </ul>
            </div>

            <div style="text-align: center; margin: 30px 0;">
                <a href="{signup_link}"
                   style="background: #5B8DEF;
                          color: #ffffff;
                          padding: 15px 40px;
                          text-decoration: none;
                          border-radius: 8px;
                          font-weight: bold;
                          display: inline-block;
                          font-size: 16px;
                          box-shadow: 0 4px 6px rgba(91, 141, 239, 0.3);">
                    Aceitar Convite e Criar Conta
                </a>
            </div>

            <p style="font-size: 14px; color: #666; margin-top: 30px;">
                Ou copie e cole este link no seu navegador:
            </p>
            <p style="font-size: 12px; color: #5B8DEF; word-break: break-all; background: white; padding: 10px; border-radius: 5px;">
                {signup_link}
            </p>

            {f'''
            <div style="background: white; padding: 20px; border-radius: 5px; margin: 30px 0; border-left: 4px solid #5B8DEF;">
                <h3 style="margin-top: 0; color: #5B8DEF;">🔑 Suas credenciais temporárias:</h3>
                <p style="margin: 10px 0;"><strong>Email:</strong> {email}</p>
                <p style="margin: 10px 0;"><strong>Senha temporária:</strong> <code style="background: #f0f0f0; padding: 4px 8px; border-radius: 4px; font-family: monospace;">{temp_password}</code></p>
                <p style="font-size: 12px; color: #666; margin-top: 15px;">
                    💡 Recomendamos alterar sua senha após o primeiro login.
                </p>
            </div>
            ''' if temp_password else ''}

            <p style="font-size: 14px; color: #666; margin-top: 30px; padding-top: 20px; border-top: 1px solid #ddd;">
                ⏰ Este convite expira em <strong>7 dias</strong>.<br>
                ⚠️ Se você não esperava este convite, pode ignorar este email com segurança.
            </p>
        </div>

        <div style="text-align: center; margin-top: 20px; color: #999; font-size: 12px;">
            <p>© {datetime.now().year} XTYL Creativity Machine. Todos os direitos reservados.</p>
        </div>
    </body>
    </html>
    """

    return send_email(
        to_email=email,
        subject=f"✉️ {invited_by} convidou você para o workspace '{workspace_name}' - XTYL",
        html_content=html_content
    )


def send_welcome_email(email: str, workspace_name: str, invited_by: str, user_name: Optional[str] = None) -> bool:
    """
    Send welcome email when user is added to a workspace

    Args:
        email: User's email address
        workspace_name: Name of the workspace
        invited_by: Name of the person who invited
        user_name: User's name (optional)

    Returns:
        bool: True if email was sent successfully
    """
    login_link = f"{FRONTEND_URL}/login"

    html_content = f"""
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Bem-vindo ao XTYL</title>
    </head>
    <body style="font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif; line-height: 1.6; color: #333; max-width: 600px; margin: 0 auto; padding: 20px;">
        <div style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); padding: 30px; text-align: center; border-radius: 10px 10px 0 0;">
            <h1 style="color: white; margin: 0; font-size: 28px;">🎉 Bem-vindo ao XTYL!</h1>
        </div>

        <div style="background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px;">
            <p style="font-size: 16px; margin-bottom: 20px;">
                Olá{f" {user_name}" if user_name else ""},
            </p>

            <p style="font-size: 16px; margin-bottom: 20px;">
                Você foi adicionado ao workspace <strong>"{workspace_name}"</strong> por <strong>{invited_by}</strong>.
            </p>

            <p style="font-size: 16px; margin-bottom: 30px;">
                O <strong>XTYL Creativity Machine</strong> é sua plataforma completa para criação e gerenciamento de conteúdo criativo com inteligência artificial.
            </p>

            <div style="background: white; padding: 20px; border-radius: 5px; margin: 20px 0;">
                <h3 style="margin-top: 0; color: #667eea;">✨ O que você pode fazer:</h3>
                <ul style="padding-left: 20px;">
                    <li style="margin-bottom: 10px;">📝 Criar e gerenciar documentos com IA</li>
                    <li style="margin-bottom: 10px;">🎨 Gerar imagens com modelos avançados</li>
                    <li style="margin-bottom: 10px;">💬 Chat inteligente com contexto</li>
                    <li style="margin-bottom: 10px;">📊 Organizar projetos em pastas</li>
                    <li style="margin-bottom: 10px;">🖼️ Biblioteca de assets visuais</li>
                </ul>
            </div>

            <div style="text-align: center; margin: 30px 0;">
                <a href="{login_link}"
                   style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
                          color: white;
                          padding: 15px 40px;
                          text-decoration: none;
                          border-radius: 5px;
                          font-weight: bold;
                          display: inline-block;
                          font-size: 16px;">
                    Acessar Plataforma
                </a>
            </div>

            <p style="font-size: 14px; color: #666; margin-top: 30px;">
                <strong>Suas credenciais:</strong><br>
                Email: {email}<br>
                Senha: (use a senha que você cadastrou ou solicite recuperação se necessário)
            </p>

            <p style="font-size: 14px; color: #666; margin-top: 30px; padding-top: 20px; border-top: 1px solid #ddd;">
                💡 <strong>Dica:</strong> Explore os templates disponíveis para começar rapidamente!
            </p>
        </div>

        <div style="text-align: center; margin-top: 20px; color: #999; font-size: 12px;">
            <p>© {datetime.now().year} XTYL Creativity Machine. Todos os direitos reservados.</p>
            <p>Se você recebeu este email por engano, por favor ignore.</p>
        </div>
    </body>
    </html>
    """

    return send_email(
        to_email=email,
        subject=f"🎉 Você foi adicionado ao workspace '{workspace_name}' - XTYL",
        html_content=html_content,
        to_name=user_name
    )

File: alvobot-reference/backend/src/modules/email/email.service.ts
import { Injectable, Logger } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { Resend } from "resend";

export interface SendEmailOptions {
  to: string | string[];
  subject: string;
  htmlContent: string;
  textContent?: string;
  replyTo?: string;
  tags?: string[];
}

export interface WorkspaceInviteEmailData {
  inviteeEmail: string;
  inviterName: string;
  workspaceName: string;
  role: string;
  inviteLink: string;
  expiresIn: string;
}

@Injectable()
export class EmailService {
  private readonly logger = new Logger(EmailService.name);
  private readonly resend: Resend | null;
  private readonly senderEmail: string;
  private readonly senderName: string;
  private readonly isConfigured: boolean;

  constructor(private configService: ConfigService) {
    const apiKey = this.configService.get<string>("RESEND_API_KEY");
    this.senderEmail =
      this.configService.get<string>("RESEND_SENDER_EMAIL") ||
      "noreply@alvobot.com";
    this.senderName =
      this.configService.get<string>("RESEND_SENDER_NAME") || "AlvoBot";

    if (apiKey) {
      this.resend = new Resend(apiKey);
      this.isConfigured = true;
      this.logger.log("Resend email service configured");
    } else {
      this.resend = null;
      this.isConfigured = false;
      this.logger.warn(
        "RESEND_API_KEY not configured - emails will be logged only",
      );
    }
  }

  /**
   * Send a transactional email via Resend
   */
  async sendEmail(options: SendEmailOptions): Promise<boolean> {
    const recipients = Array.isArray(options.to) ? options.to : [options.to];

    if (!this.isConfigured || !this.resend) {
      this.logger.warn(
        `[EMAIL NOT SENT - NO API KEY] To: ${recipients.join(", ")}, Subject: ${options.subject}`,
      );
      this.logger.debug(
        `HTML Content: ${options.htmlContent.substring(0, 500)}...`,
      );
      return false;
    }

    try {
      const { data, error } = await this.resend.emails.send({
        from: `${this.senderName} <${this.senderEmail}>`,
        to: recipients,
        subject: options.subject,
        html: options.htmlContent,
        text: options.textContent,
        replyTo: options.replyTo,
        tags: options.tags?.map((tag) => ({ name: tag, value: "true" })),
      });

      if (error) {
        this.logger.error(`Failed to send email: ${error.message}`, error);
        return false;
      }

      this.logger.log(
        `Email sent successfully to ${recipients.join(", ")} (ID: ${data?.id})`,
      );
      return true;
    } catch (error) {
      this.logger.error(`Failed to send email: ${error.message}`, error.stack);
      return false;
    }
  }

  /**
   * Send workspace invitation email
   */
  async sendWorkspaceInvitation(
    data: WorkspaceInviteEmailData,
  ): Promise<boolean> {
    const roleLabels: Record<string, string> = {
      admin: "Administrador",
      member: "Membro",
      viewer: "Visualizador",
    };

    const roleLabel = roleLabels[data.role] || data.role;

    const htmlContent = this.getWorkspaceInviteTemplate(data, roleLabel);
    const textContent = this.getWorkspaceInviteTextTemplate(data, roleLabel);

    return this.sendEmail({
      to: data.inviteeEmail,
      subject: `${data.inviterName} convidou você para o workspace "${data.workspaceName}"`,
      htmlContent,
      textContent,
      tags: ["workspace-invite"],
    });
  }

  /**
   * HTML template for workspace invitation
   */
  private getWorkspaceInviteTemplate(
    data: WorkspaceInviteEmailData,
    roleLabel: string,
  ): string {
    return `
<!DOCTYPE html>
<html lang="pt-BR">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Convite para Workspace</title>
</head>
<body style="margin: 0; padding: 0; font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, sans-serif; background-color: #f5f5f5;">
  <table role="presentation" width="100%" cellspacing="0" cellpadding="0" style="background-color: #f5f5f5; padding: 40px 20px;">
    <tr>
      <td align="center">
        <table role="presentation" width="600" cellspacing="0" cellpadding="0" style="background-color: #ffffff; border-radius: 12px; box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);">
          <!-- Header -->
          <tr>
            <td style="padding: 40px 40px 20px; text-align: center; background: linear-gradient(135deg, #fbbf24 0%, #f59e0b 100%); border-radius: 12px 12px 0 0;">
              <h1 style="margin: 0; color: #1a1a2e; font-size: 28px; font-weight: 700;">AlvoBot</h1>
            </td>
          </tr>

          <!-- Content -->
          <tr>
            <td style="padding: 40px;">
              <h2 style="margin: 0 0 16px; color: #1a1a2e; font-size: 24px; font-weight: 600;">
                Você foi convidado! 🎉
              </h2>

              <p style="margin: 0 0 24px; color: #4a5568; font-size: 16px; line-height: 1.6;">
                <strong>${data.inviterName}</strong> convidou você para participar do workspace
                <strong>"${data.workspaceName}"</strong> como <strong>${roleLabel}</strong>.
              </p>

              <div style="background-color: #f8fafc; border-radius: 8px; padding: 20px; margin-bottom: 24px;">
                <table role="presentation" width="100%" cellspacing="0" cellpadding="0">
                  <tr>
                    <td style="padding: 8px 0;">
                      <span style="color: #718096; font-size: 14px;">Workspace:</span>
                      <span style="color: #1a1a2e; font-size: 14px; font-weight: 600; float: right;">${data.workspaceName}</span>
                    </td>
                  </tr>
                  <tr>
                    <td style="padding: 8px 0; border-top: 1px solid #e2e8f0;">
                      <span style="color: #718096; font-size: 14px;">Seu papel:</span>
                      <span style="color: #1a1a2e; font-size: 14px; font-weight: 600; float: right;">${roleLabel}</span>
                    </td>
                  </tr>
                  <tr>
                    <td style="padding: 8px 0; border-top: 1px solid #e2e8f0;">
                      <span style="color: #718096; font-size: 14px;">Convite expira em:</span>
                      <span style="color: #e53e3e; font-size: 14px; font-weight: 600; float: right;">${data.expiresIn}</span>
                    </td>
                  </tr>
                </table>
              </div>

              <!-- CTA Button -->
              <table role="presentation" width="100%" cellspacing="0" cellpadding="0">
                <tr>
                  <td align="center" style="padding: 16px 0;">
                    <a href="${data.inviteLink}"
                       style="display: inline-block; background-color: #fbbf24; color: #1a1a2e; text-decoration: none; font-weight: 600; font-size: 16px; padding: 14px 32px; border-radius: 8px; box-shadow: 0 2px 4px rgba(251, 191, 36, 0.3);">
                      Aceitar Convite
                    </a>
                  </td>
                </tr>
              </table>

              <p style="margin: 24px 0 0; color: #718096; font-size: 14px; text-align: center;">
                Se o botão não funcionar, copie e cole este link no seu navegador:
              </p>
              <p style="margin: 8px 0 0; color: #4a5568; font-size: 12px; word-break: break-all; text-align: center;">
                <a href="${data.inviteLink}" style="color: #fbbf24;">${data.inviteLink}</a>
              </p>
            </td>
          </tr>

          <!-- Footer -->
          <tr>
            <td style="padding: 24px 40px; background-color: #f8fafc; border-radius: 0 0 12px 12px; text-align: center;">
              <p style="margin: 0; color: #718096; font-size: 12px;">
                Este email foi enviado porque alguém convidou você para um workspace no AlvoBot.
                <br>Se você não reconhece este convite, pode ignorar este email com segurança.
              </p>
              <p style="margin: 16px 0 0; color: #a0aec0; font-size: 11px;">
                © ${new Date().getFullYear()} AlvoBot. Todos os direitos reservados.
              </p>
            </td>
          </tr>
        </table>
      </td>
    </tr>
  </table>
</body>
</html>
`;
  }

  /**
   * Plain text template for workspace invitation
   */
  private getWorkspaceInviteTextTemplate(
    data: WorkspaceInviteEmailData,
    roleLabel: string,
  ): string {
    return `
Você foi convidado para o AlvoBot!

${data.inviterName} convidou você para participar do workspace "${data.workspaceName}" como ${roleLabel}.

Detalhes do convite:
- Workspace: ${data.workspaceName}
- Seu papel: ${roleLabel}
- Expira em: ${data.expiresIn}

Para aceitar o convite, acesse o link abaixo:
${data.inviteLink}

Se você não reconhece este convite, pode ignorar este email com segurança.

---
AlvoBot - ${new Date().getFullYear()}
`.trim();
  }
}
